File: backend/test_before_commit.py
#!/usr/bin/env python3
"""
Pre-commit test runner for Buzz2Remote backend.
Runs essential tests and blocks deployment if they fail.
Priority: Syntax → Unit → API → Integration
"""

import subprocess
import sys
import os
from datetime import datetime

def run_command(cmd):
    """Run a command and return success status."""
    try:
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
        return result.returncode == 0, result.stdout, result.stderr
    except Exception as e:
        return False, "", str(e)

def main():
    """Main test runner."""
    print(f"""
🧪 Pre-Commit Test Suite
Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}
Priority: Syntax → Unit → API → Integration
""")
    
    # Tests to run in priority order (most critical first)
    tests = [
        ("🔍 Syntax & Import Tests", "pytest tests/syntax/ -q"),
        ("🔧 Unit Tests", "pytest tests/unit/ -q"),
        ("🌐 Basic API Tests", "pytest tests/api/test_health.py -q"),
        ("🔗 Database Tests", "pytest tests/unit/test_database.py -q"),
    ]
    
    all_passed = True
    
    for test_name, cmd in tests:
        print(f"Running {test_name}...")
        success, stdout, stderr = run_command(cmd)
        
        if success:
            print(f"✅ {test_name} PASSED")
        else:
            print(f"❌ {test_name} FAILED")
            print(f"Error: {stderr}")
            all_passed = False
            # Stop at first failure for faster feedback
            break
    
    print("\n" + "="*50)
    
    if all_passed:
        print("🎉 ALL TESTS PASSED! Safe to deploy.")
        print("💡 Run 'pytest --cov=.' for full test suite with coverage")
        return 0
    else:
        print("💥 TESTS FAILED! DO NOT DEPLOY!")
        print("🔧 Fix the failing tests before committing.")
        print("📝 Most likely issues:")
        print("   - Syntax errors (indentation, missing colons)")
        print("   - Import errors (missing dependencies)")
        print("   - Basic logic errors")
        return 1

if __name__ == "__main__":
    exit_code = main()
    sys.exit(exit_code) 

File: frontend/test-before-commit.js
#!/usr/bin/env node

/**
 * Pre-commit test runner for Buzz2Remote frontend.
 * Runs essential tests and blocks deployment if they fail.
 * Priority: Syntax → Unit → Integration
 */

const { spawn } = require('child_process');

function runCommand(command, args = []) {
  return new Promise((resolve) => {
    const process = spawn(command, args, { 
      stdio: 'pipe',
      shell: true 
    });

    let stdout = '';
    let stderr = '';

    process.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    process.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    process.on('close', (code) => {
      resolve({
        success: code === 0,
        stdout,
        stderr,
        code
      });
    });
  });
}

async function main() {
  console.log(`
🧪 Frontend Pre-Commit Test Suite
Time: ${new Date().toLocaleString()}
Priority: Syntax → Unit → Build → Lint
`);

  // Tests to run in priority order (most critical first)
  const tests = [
    {
      name: "🔍 Build Syntax & Import Tests Structure Tests",
      command: "npm test -- --testPathPattern=build-test --watchAll=false --verbose=false"
    },
    {
      name: "🔧 Unit Tests", 
      command: "npm test -- --testPathPattern=unit --watchAll=false --verbose=false"
    },
    {
      name: "🏗️ Build Test",
      command: "npm run build"
    },
    {
      name: "📋 Lint Check",
      command: "npm run lint"
    }
  ];

  let allPassed = true;

  for (const test of tests) {
    console.log(`Running ${test.name}...`);
    
    const result = await runCommand(test.command);
    
    if (result.success) {
      console.log(`✅ ${test.name} PASSED`);
    } else {
      console.log(`❌ ${test.name} FAILED`);
      console.log(`Error: ${result.stderr}`);
      
      // Only show critical errors, not all build warnings
      if (result.stderr && !result.stderr.includes('Warning:')) {
        console.log(`Details: ${result.stdout.substring(0, 500)}...`);
      }
      
      allPassed = false;
      // Stop at first failure for faster feedback
      break;
    }
  }

  console.log("\n" + "=".repeat(50));

  if (allPassed) {
    console.log("🎉 ALL FRONTEND TESTS PASSED! Safe to deploy.");
    console.log("💡 Run 'npm test -- --coverage' for full test suite with coverage");
    process.exit(0);
  } else {
    console.log("💥 FRONTEND TESTS FAILED! DO NOT DEPLOY!");
    console.log("🔧 Fix the failing tests before committing.");
    console.log("📝 Most likely issues:");
    console.log("   - Syntax errors (missing semicolons, brackets)");
    console.log("   - Import errors (missing dependencies, wrong paths)");
    console.log("   - Build errors (TypeScript, React compilation)");
    console.log("   - Component rendering issues");
    process.exit(1);
  }
}

// Handle process termination gracefully
process.on('SIGINT', () => {
  console.log('\n🛑 Test run interrupted by user');
  process.exit(1);
});

process.on('SIGTERM', () => {
  console.log('\n🛑 Test run terminated');
  process.exit(1);
});

main().catch(error => {
  console.error('❌ Test runner failed:', error.message);
  process.exit(1);
}); 